// @here/hds-vertical-navigation/build/hds-vertical-navigation-item.js
var VerticalNavigationItem_1;
import { __decorate } from "tslib";
import { customElement } from '@here/hds-base';
import { AccordionItem } from '@here/hds-accordion';
import '@here/hds-avatar';
import { html, nothing } from 'lit';
import { property, query } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { styles } from './hds-vertical-navigation-item.css.js';
import { VerticalNavigationSubList } from './hds-vertical-navigation-sub-list.js';
/**
 * @slot default Content for a `hds-vertical-navigation-sub-list` structure
 * @event {CustomEvent<{verticalNavItemId: string}>} select Fired when the item has been selected `VerticalNavigationSelectPayload`
 */
let VerticalNavigationItem = VerticalNavigationItem_1 = class VerticalNavigationItem extends AccordionItem {
    constructor() {
        super(...arguments);
        this.hasSubList = false;
        this.chevron = false;
        this.navMode = 'expanded';
        this.avatarUrl = '';
        this.avatarName = '';
        this.iconStyle = 'solid';
        this.selected = false;
    }
    static get styles() {
        return [super.styles, styles];
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.id = this.id ?? `hds-vertical-navigation-item-${++VerticalNavigationItem_1.itemIdCounter}`;
    }
    _slotChange() {
        let hasSubList = false;
        this.slotElement.assignedElements().forEach(e => {
            if (e instanceof VerticalNavigationSubList) {
                hasSubList = true;
            }
            else {
                e.slot = 'headline';
            }
        });
        this.hasSubList = hasSubList;
    }
    updated(_changedProperties) {
        super.updated(_changedProperties);
        if (!this.disabled && this.selected && _changedProperties.has('selected')) {
            this.dispatchEvent(new CustomEvent('select', {
                detail: { verticalNavItemId: this.id, verticalNavItemIndex: this.index },
                bubbles: true,
                composed: true,
            }));
        }
    }
    renderIcon() {
        let icon = nothing;
        if (this.icon) {
            icon = html `
        <hds-icon
          class="icon"
          name="${this.icon}"
          category="${this.iconCategory}"
          icon-style="${this.iconStyle}"
          size="24px"
        ></hds-icon>
      `;
        }
        else if (this.avatarUrl || this.avatarName) {
            icon = html ` <hds-avatar
        size="medium"
        name="${ifDefined(this.avatarName)}"
        image-url="${ifDefined(this.avatarUrl)}"
      ></hds-avatar>`;
        }
        return icon;
    }
    render() {
        const classes = {
            'has-sub-list': this.hasSubList,
            'has-chevron': this.chevron,
            'has-avatar': this.avatarUrl || this.avatarName,
            'nav-expanded': this.navMode === 'expanded',
            'nav-collapsed': this.navMode === 'collapsed',
            selected: this.selected,
        };
        return html `
      <div
        class="-hds-vertical-navigation-item ${classMap(classes)}"
        aria-selected="${ifDefined(this.selected)}"
        tabindex="0"
        @slotchange="${this._slotChange}"
        @click="${this._handleClick}"
      >
        ${super.render()}
      </div>
    `;
    }
    async _handleClick(e) {
        e.stopPropagation();
        if (!this.disabled)
            this.selected = true;
    }
};
VerticalNavigationItem.itemIdCounter = 0;
__decorate([
    property({ type: Boolean, attribute: false })
], VerticalNavigationItem.prototype, "hasSubList", void 0);
__decorate([
    property({ type: Boolean, attribute: false })
], VerticalNavigationItem.prototype, "chevron", void 0);
__decorate([
    property({ type: String, attribute: false })
], VerticalNavigationItem.prototype, "navMode", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'avatar-url' })
], VerticalNavigationItem.prototype, "avatarUrl", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'avatar-name' })
], VerticalNavigationItem.prototype, "avatarName", void 0);
__decorate([
    property({ type: String, attribute: 'icon-style' })
], VerticalNavigationItem.prototype, "iconStyle", void 0);
__decorate([
    property({ type: Boolean, attribute: true, reflect: true })
], VerticalNavigationItem.prototype, "selected", void 0);
__decorate([
    query('slot:not([name])')
], VerticalNavigationItem.prototype, "slotElement", void 0);
VerticalNavigationItem = VerticalNavigationItem_1 = __decorate([
    customElement('hds-vertical-navigation-item')
], VerticalNavigationItem);
export { VerticalNavigationItem };
