// @here/hds-vertical-navigation/build/hds-vertical-navigation-sub-list.js
var VerticalNavigationSubList_1;
import { __decorate } from "tslib";
import { customElement } from '@here/hds-base';
import { List } from '@here/hds-list';
import { html } from 'lit';
import { property, query } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { VerticalNavigationSubListItem } from './hds-vertical-navigation-sub-list-item.js';
import { styles } from './hds-vertical-navigation-sub-list.css.js';
/**
 * @slot default Content for the list of `hds-vertical-navigation-sub-list-item` elements
 */
let VerticalNavigationSubList = VerticalNavigationSubList_1 = class VerticalNavigationSubList extends List {
    constructor() {
        super(...arguments);
        this.isMainSubList = false;
    }
    static get styles() {
        return [super.styles, styles];
    }
    updated(_changedProperties) {
        super.updated(_changedProperties);
    }
    firstUpdated() {
        this._slotChange();
    }
    _slotChange() {
        this.isMainSubList =
            !(this.parentNode instanceof VerticalNavigationSubList_1) &&
                !(this.parentNode instanceof VerticalNavigationSubListItem);
    }
    render() {
        const classes = {
            '-hds-vertical-navigation-sub-list': true,
            'main-sub-list': this.isMainSubList,
        };
        return html `
      <div class="${classMap(classes)}" @slotchange="${this._slotChange}">${super.render()}</div>
    `;
    }
};
__decorate([
    property({ type: Boolean, attribute: false })
], VerticalNavigationSubList.prototype, "isMainSubList", void 0);
__decorate([
    query('slot:not([name])')
], VerticalNavigationSubList.prototype, "slotElement", void 0);
VerticalNavigationSubList = VerticalNavigationSubList_1 = __decorate([
    customElement('hds-vertical-navigation-sub-list')
], VerticalNavigationSubList);
export { VerticalNavigationSubList };
