// @here/hds-vertical-navigation/build/hds-vertical-navigation-sub-list-item.js
import { __decorate } from "tslib";
import { customElement, findElementsInSlot } from '@here/hds-base';
import { ListItem } from '@here/hds-list';
import { html } from 'lit';
import { property, query } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { styles } from './hds-vertical-navigation-sub-list-item.css.js';
import { VerticalNavigationSubList } from './hds-vertical-navigation-sub-list.js';
/**
 * @slot default Content for the item's label
 *
 * @event {CustomEvent<{verticalSubListItemId: string}>}  subListItemSelect (**Internal event**) Fired when item has been selected, internal use only `VerticalNavigationSelectPayload`
 */
let VerticalNavigationSubListItem = class VerticalNavigationSubListItem extends ListItem {
    constructor() {
        super(...arguments);
        this.hasSubList = false;
        this._assignedElementsCache = [];
    }
    static get styles() {
        return [super.styles, styles];
    }
    firstUpdated() {
        super.firstUpdated();
        this.hasTooltip = false;
        this._slotChange();
    }
    async updated(changedProperties) {
        this._updateAssignedElements();
        if (changedProperties.has('selected')) {
            if (this.selected && !this.disabled) {
                this.dispatchEvent(new CustomEvent('subListItemSelect', {
                    detail: { verticalSubListItemId: this.id },
                    bubbles: true,
                    composed: true,
                }));
            }
        }
    }
    _slotChange() {
        this._assignedElementsCache = findElementsInSlot(this.slotElement, VerticalNavigationSubList);
        this._assignedElementsCache.forEach(e => {
            if (e instanceof VerticalNavigationSubList) {
                e.slot = 'sub-list';
            }
        });
        this._updateAssignedElements();
    }
    _updateAssignedElements() {
        this._assignedElementsCache.forEach(e => {
            if (e instanceof VerticalNavigationSubList) {
                e.disabled = this.disabled;
            }
        });
    }
    render() {
        const classes = {
            disabled: this.disabled || this.parentDisabled,
        };
        return html `
      <div
        class="-hds-vertical-navigation-sub-list-item ${classMap(classes)}"
        @click="${this._handleClick}"
        @slotchange="${this._slotChange}"
      >
        <div class="main-list-wrapper">${super.render()}</div>
        <div class="sub-list-wrapper">
          <slot name="sub-list"></slot>
        </div>
      </div>
    `;
    }
    _handleClick(e) {
        e.stopPropagation();
    }
};
__decorate([
    property({ type: Boolean, attribute: false, reflect: true })
], VerticalNavigationSubListItem.prototype, "hasSubList", void 0);
__decorate([
    query('.sub-list-wrapper')
], VerticalNavigationSubListItem.prototype, "subListElement", void 0);
__decorate([
    query('slot:not([name])')
], VerticalNavigationSubListItem.prototype, "slotElement", void 0);
VerticalNavigationSubListItem = __decorate([
    customElement('hds-vertical-navigation-sub-list-item')
], VerticalNavigationSubListItem);
export { VerticalNavigationSubListItem };
